File: dotnet/Allmark/Inline/SubscriptRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class SubscriptRule
{
    public static InlineRule Create()
    {
        return new InlineRule
        {
            Name = "subscript",
            Test = TestSubscript,
            Precedence = 5,
        };
    }

    private static bool TestSubscript(InlineParserState state, MarkdownNode parent)
    {
        var ch = Utils.GetChar(state.Src, state.I);
        if (!state.IsEscaped && ch == '~')
        {
            // Subscripts can only be one character long, otherwise they are a GFM strikethrough
            if (Utils.GetChar(state.Src, state.I + 1) == '~')
            {
                return false;
            }
            var rule = Create();
            return TagMarksRule.Execute("subscript", ch.ToString(), state, parent, rule.Precedence!.Value);
        }
        return false;
    }
}

File: zig/src/inline/subscriptRule.zig
const std = @import("std");
const InlineParserState = @import("../types/InlineParserState.zig").InlineParserState;
const InlineRule = @import("../types/InlineRule.zig").InlineRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const testTagMarks = @import("tagMarksRule.zig").testTagMarks;

pub fn testSubscript(state: *InlineParserState, parent: *MarkdownNode) bool {
    if (!state.isEscaped and state.i < state.src.len and state.src[state.i] == '~') {
        if (state.i + 1 < state.src.len and state.src[state.i + 1] == '~') {
            return false;
        }
        return testTagMarks("subscript", '~', state, parent, subscriptRule.precedence.?);
    }
    return false;
}

pub const subscriptRule = InlineRule{
    .name = "subscript",
    .@"test" = testSubscript,
    .precedence = 5,
};
